### ISSUE TITLE
CPUDevice->Allocate for large-tr temporal filtering

### ISSUE THREAD

**Issue by DTL2020** [AviSynth/AviSynthPlus#351]
Currently CPUDevice->Allocat() uses simple new method of memory allocation. 

It is https://github.com/AviSynth/AviSynthPlus/blob/45404659f0fed07e369978924efd417221bdc782/avs_core/core/DeviceManager.cpp#L180 

It is not guaranteed to be SIMD aligned for SSE/AVX/AVX512 ? But also it is not guaranteed to be safe for long-tr temporal processing filters with many sequential frames parallel access for reading. If the reading addresses wil overload number of ways in cache set it will lost performance significantly. Typical number of ways in cache sets of L1 is about 8 so with tr>4 it is already possible performance penalty.

2 possible solutions exist:
1. Fast but poorer in non-supervision controllable memory allocation - allocate more RAM in some number of cachelines max and add random offset of cacheline sized to each start address. It not require tracking of all used start address offsets but can hit same used addresses because random number generator can return equal numbers. Also some address space is poorerly used (maybe more RAM waisted).
It is solution used in large-page mod - https://github.com/DTL2020/AviSynthPlus-LP-mod/blob/62d7c268f279d4f6917b323e6806082847603b67/DeviceManager.cpp#L198

3. Make some global class in AVS core to watch for the sequential frames starting addresses allocations and simply advance with cacheline size each next frame starting address allocation and reset addition value to zero  after all required addresses are used (the total max offset can be calculated from current CPU cache params like number of ways and cache size and cacheline size). So all sequential frames in AVS environment will have guaranteed offset to hit different cache sets and the probability of performance penalty for large-tr temporal filters will be probably removed completely.

**Comment by pinterf** [comment]
The place you found is not where the allocation rules are ensured. It just allocates the given number of bytes (and a bit more for partly debug ('deadbeef', but yes, the +16 bytes are there in non-debug builds for some reason) and cuda ('margin') reasons)

Avisynth plus aligns pitch and frame pointers at 64 byte boundaries. I changed this at the very beginning of my involvement in the project.

This is done by setting a frame_align to 64 from FRAME_ALIGN, which is defined as 64 in avisynth.h.

When Nekopanda (in Neo branch) fixed a lot of MT issues and introduced the possibility of a CUDA-only pipeline, then it was changed not to be constant, at least not for a non-CPU device. The 'frame_align'  and 'plane_align' variables was introduced then.
Allocation rules for CUDA are different, 64 bytes alignment may be too much.

https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/core/avisynth.cpp#L2346

The actual alignment in a video frame, setting the beginning offset and the pitches are all done in NewVideoFrame
https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/core/avisynth.cpp#L3576
or
https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/core/avisynth.cpp#L3636

**Comment by DTL2020** [comment]
Yes - I see the 64-bytes alignment is performed. But I still not see if the required offsets in virtual address space is added for different sequential frames placement in memory to save from possible cache aliasing conditions if some processing filter do large-tr frames access (i.e. read same x,y coordinate sample or SIMD-dataword group of samples of cacheline sized 32 or 64 bytes for AVX/AVX512 from a sequence of frames +-tr from current frame).

In my experience with LargePage allocations and not adding offset for different frames allocations the performance of mvtools/MDegrainN with tr about 10 decrease at about 50% (the most suffering is MDegerainN processing because MAnalyse only process pairs of frames and MDegrainN access all tr-pooled frames in parallel). Now we also have vsTTempSmooth sample-based temporal processing filter with also tr of several dozens in good case (or equal to mvtools MAX_TRAD of 128).

With random start addressing for 'new' method the performance may be somehow inbetween the worst case (all frames/rows start addresses aligned in worst case for hitting same cache set) and the best (the sets of cache filled sequentially in best order).

Some calculations on typical 32 kB L1D cache and 8-ways with 64bytes cacheline size:

The cache architecture typically designed to be completely filled with sequential memory access (read for example). So memcpy of _read_all_cache and _write_all_cache is sort of best performance to save from bus directions switching.

It mean sequential read of 32 kBytes of virtual addresses fill all sets of all cache. So with 64 bytes of cacheline and 8-ways - size of each set is 64x8=512bytes. Number of sets is 32768/512=64 sets. Each set can serve only 8 aliasing memory addresses. Also it may be no good to fill all set because some other process data from unknown addresses also need to be cached. So it may be good to fill only Nways/2 in each set.

So the equal x,y coordinates of sequential frames allocations in AVS core expected to have virtual addresses that fill (hit) only about Nways/2 of each cache set when some large-tr filter process frames in 'parallel access' mode.

**Comment by pinterf** [comment]
I suppose this one
https://github.com/DTL2020/AviSynthPlus-LP-mod/blob/62d7c268f279d4f6917b323e6806082847603b67/DeviceManager.cpp#L275
works only if granularity is exactly 2M?
https://github.com/DTL2020/AviSynthPlus-LP-mod/blob/62d7c268f279d4f6917b323e6806082847603b67/DeviceManager.cpp#L249

Does it really need to use VirtualAlloc?
Can it work when just using the memory allocation of the actual C++ library and provide the random offset back?

I think I slowly understand the 4k aliasing problem,
This is also a good link, from which I noticed that the processing can really suffer a lot from this phenomenon.
https://stackoverflow.com/questions/25774190/l1-memory-bandwidth-50-drop-in-efficiency-using-addresses-which-differ-by-4096

Aside from the fact that specific scripts such as MDegrainN over +/-5 or more frames would benefit, doesn't it hurt other use cases? Large pages support alone can give surprises, I read come considetations here as well: https://reviews.llvm.org/D58718
 (And we are talking only about tuning an Avs+ running on Windows 10-11 with 16GiB+ memory?)

**Comment by DTL2020** [comment]
"https://github.com/DTL2020/AviSynthPlus-LP-mod/blob/62d7c268f279d4f6917b323e6806082847603b67/DeviceManager.cpp#L275
works only if granularity is exactly 2M?"

That method of ptr = (BYTE*)((uint64_t)ptr >> 21); is simply to not store separately pointer to free memory. In better case programmer need to store several pointers to each memory allocation - the pointer to free memory region and working pointer for application. I not sure if VirtualFree with some 'random internal address of allocated memory block' will free memory region correctly.

"Does it really need to use VirtualAlloc?"

From programmer's view the use of OS-dependent VirtualAlloc is not strictly required. It is some handy method in Windows to get page-aligned pointer and also zeroed memory region (it is physically zeroed by some Windows service threads in idle system time - so returned by VirtualAlloc memory region already guaranteed to be zeroed-clean). But with C-standard (OS-independent) memory allocation it is not known starting virtual address of allocation so additional memory area size padding to play with real working address range inside allocation may be twice more in size (?).

"Can it work when just using the memory allocation of the actual C++ library and provide the random offset back?"

Yes - it can. The only issue may be a bit more RAM waisted for larger padding to have enough space to play with the allocated region for placing actual working buffer (start and end virtual addresses).

"the 4k aliasing problem,"

The aliasing problem is not only 4K - the 4K is only included in all possible aliasing cases. The best solution is to keep track of actual virtual memory to cache sets mapping and sort of having 'watch algorithm' keeping usage (filling) of cache sets at level around 50% of total used ways in each set. It is typical recommendation of intel optimizing - to optimize to about 50% of cache size.

" doesn't it hurt other use cases?"

It is expected also with vsTTempSmooth and any other 'large-tr' temporal filter accessing 'too much' frames around current in 'parallel-way' (same x,y coordinates reading).

"Large pages support alone can give surprises,"

LP unlikely usable in current AVS without much more core redesign - the typical windows only reserve around 256 MBs only for LP allocations after fresh system reboot. So it is required to design OS-startup driver to allocate required GBs of contigous RAM (using file or system registry record how much to allocate) and provide to AVS core (via DeviceIoCtl call) after system complete bootup. Also this RAM will be lost form all other Windows and processes. It is special AVS use case.

"we are talking only about tuning an Avs+ running on Windows 10-11 with 16GiB+ memory?"

Set-assiciative cache aliasing issues exist at any RAM size for Windows bootup. Like from 1 MB in size. Also it is CPU hardware issue - not OS-dependent.

**Comment by DTL2020** [comment]
" good link, from which I noticed that the processing can really suffer a lot from this phenomenon.
https://stackoverflow.com/questions/25774190/l1-memory-bandwidth-50-drop-in-efficiency-using-addresses-which-differ-by-4096"

That issue may be a bit differently if only access 1 reading location and 1 writing. It may be about described in intel manual about read-write with some bad stride: Store forwarding — If a store to an address is followed by a load from the same address, the load
will not proceed until the store data is available. If a store is followed by a load and their addresses
differ by a multiple of 4 KBytes, the load stalls until the store operation completes.

**Comment by DTL2020** [comment]
Sort of short description of issue:
For set-associative cache computing platforms many different ranges of virtual addresses of total virtual address space mapped to single cache set. Even if all resources of single cache set are busy (full) - the mapping can not be changed in runtime to other free cache sets (it will degrade chip performance for cache search algorithm). It is soft of short description of 'cache aliasing issue'.

Programming task for memory manager designer for many read/(write) streams processing environments: adjust virtual addresses of different read/write streams so they will map to different cache sets. So each read/write stream of many parallel streams can be served by individual cache set and total performance of memory subsystem is maximum.
